gematria/datasets/exegesis_benchmark_lib_test.cc

```diff
@@ -0,0 +1,386 @@
+#include "llvm/Support/Errc.h"
+#include "llvm/Support/Error.h"
+#include "llvm/Support/JSON.h"
+#include "llvm/Support/TargetSelect.h"
+#include "llvm/lib/Target/X86/MCTargetDesc/X86MCTargetDesc.h"
+#include "llvm/tools/llvm-exegesis/lib/BenchmarkCode.h"
+#include "llvm/tools/llvm-exegesis/lib/PerfHelper.h"
+#include "llvm/tools/llvm-exegesis/lib/TargetSelect.h"
+
+namespace gematria {
+namespace {
+
+using namespace llvm;
+using namespace llvm::exegesis;
+
+class ExegesisBenchmarkTest : public testing::Test {
+ protected:
+  std::unique_ptr<ExegesisBenchmark> Benchmark;
+
+ protected:
+  ExegesisBenchmarkTest() : Benchmark(cantFail(ExegesisBenchmark::create())){};
+
+  static void SetUpTestSuite() {
+    // LLVM Setup
+    LLVMInitializeX86TargetInfo();
+    LLVMInitializeX86TargetMC();
+    LLVMInitializeX86Target();
+    LLVMInitializeX86AsmPrinter();
+    LLVMInitializeX86AsmParser();
+    LLVMInitializeX86Disassembler();
+
+    // Exegesis Setup
+    InitializeX86ExegesisTarget();
+
+    pfm::pfmInitialize();
+  }
+
+  Expected<std::string> getErrorMessage(StringRef JSONBlock) {
+    Expected<json::Value> BlockValue = json::parse(JSONBlock);
+    if (!BlockValue) return BlockValue.takeError();
+
+    Expected<BenchmarkCode> BenchCode =
+        Benchmark->parseJSONBlock(*BlockValue->getAsObject(), 1);
+    if (BenchCode)
+      return llvm::make_error<StringError>(errc::interrupted,
+                                           "Failed to get BenchmarkCode");
+
+    std::string ErrorMessage;
+    handleAllErrors(BenchCode.takeError(),
+                    [&ErrorMessage](StringError &BenchCodeError) {
+                      ErrorMessage = BenchCodeError.getMessage();
+                    });
+
+    return ErrorMessage;
+  }
+
+  Expected<double> benchmark(StringRef JSONBlock) {
+    Expected<json::Value> BlockValue = json::parse(JSONBlock);
+    if (!BlockValue) return BlockValue.takeError();
+
+    Expected<BenchmarkCode> BenchCode =
+        Benchmark->parseJSONBlock(*BlockValue->getAsObject(), 0);
+    if (!BenchCode) return BenchCode.takeError();
+
+    return Benchmark->benchmarkBasicBlock(*BenchCode);
+  }
+};
+
+TEST_F(ExegesisBenchmarkTest, TestParseJSONBlock) {
+  StringRef JSONBlock = R"json(
+  {
+    "Hex": "3b31",
+    "LoopRegister": 51,
+    "MemoryDefinitions": [
+      {
+        "Name": "MEM",
+        "Size": 4096,
+        "Value": 1
+      }
+    ],
+    "MemoryMappings": [
+      {
+        "Address": 86016,
+        "Value": "MEM"
+      }
+    ],
+    "RegisterDefinitions": [
+      {
+        "Register": 54,
+        "Value": 86016
+      },
+      {
+        "Register": 60,
+        "Value": 86016
+      }
+    ]
+  }
+  )json";
+
+  Expected<json::Value> BlockValue = json::parse(JSONBlock);
+  ASSERT_TRUE(static_cast<bool>(BlockValue));
+  Expected<BenchmarkCode> BenchCode =
+      Benchmark->parseJSONBlock(*BlockValue->getAsObject(), 1);
+
+  ASSERT_THAT(BenchCode->Key.Instructions, testing::SizeIs(1));
```

Comment:
Consider using some of the container matchers rather than asserting on the size and then pulling out elements.